### ISSUE TITLE
Tracker: cupyx.scipy.interpolate

### ISSUE THREAD

**Issue by andfoy** [cupy/cupy#7186]
### Description

This is an issue to keep track of the current improvement work to the `cupyx.scipy.interpolate` module.  The order of the checkboxes denote the planned order and priority of implementation of each module. This plan was devised by @ev-br and is subject to change and discussion.

## First phase
- [ ] `BSpline` → https://github.com/cupy/cupy/pull/7086
- [ ] `make_interp_spline`
- [ ] `PPoly`
- [ ] `CubicHermiteSpline`: Required by `pchip/akima/CubicSpline`
- [ ] `PchipInterpolator`: aliases (`pchip`, `pchip_interpolate`)
- [ ] `Akima1DInterpolator`
- [ ] `RegularGridInterpolator`: methods `linear` and `nearest`
- [ ] `RBFInterpolator`

## Second phase (details subject to change)
- [ ] `RegularGridInterpolator`: methods `cubic`, `quintic`, `slinear`, `pchip`
- [ ] `Interpn`: alias to `RegularGridInterpolator`
- [ ] `CubicBSpline`
- [ ] `splantider`, `splder`

----

##  QHull-dependent modules
- [ ] `BPoly`
- [ ] `NdPPoly`
- [ ] `make_lsq_spline`
- [ ]  `interp1d`: @ev-br states that it could be ignored (maybe?)

## "Odds and ends"
 - [x] `BarycentricInterpolator` → https://github.com/cupy/cupy/pull/6958
 - [x] `barycentric_interpolate` → https://github.com/cupy/cupy/pull/6958
 - [x] `KroghInterpolator` → https://github.com/cupy/cupy/pull/6990
 - [x] `krogh_interpolate` → https://github.com/cupy/cupy/pull/6990
 - [ ] `approximate_taylor_polynomial`
 - [ ] `lagrange` 
 - [ ] `pade`
 
### QHull-dependent
- [ ] `CloughTocher2DInterpolator
- [ ] `LinearNDInterpolator`
- [ ] `NearestNDInterpolator`
- [ ] `griddata`: a wrapper over the above three

### FITPACK code (Black-box Fortran code)
- [ ] `BivariateSpline`
- [ ] `LSQBivariateSpline`
- [ ] `LSQSphereBivariateSpline`
- [ ] `RectBivariateSpline`
- [ ] `RectSphereBivariateSpline`
- [ ] `SmoothBivariateSpline`
- [ ] `SmoothSphereBivariateSpline`
- [ ] `UnivariateSpline`
- [ ] `InterpolatedUnivariateSpline`
- [ ] `LSQUnivariateSpline`
- [ ] `bisplev`
- [ ] `bisplrep`
- [ ] `insert`
- [ ] `spalde`
- [ ] `splev`
- [ ] `splint`
- [ ] `splprep`
- [ ] `splrep`
- [ ] `sproot`

---

## Deprecated APIs
* `Interp2d`
* `Rbf`

## New APIs (not available in SciPy main)
- [ ]  `make_smooth_spline`
- [ ] `NdBSpline`

Please feel free to discuss any changes to this roadmap/exploration. 

cc @rgommers @ev-br

**Comment by kmaehashi** [comment]
Thanks for the tracker @andfoy!

**Comment by ev-br** [comment]
I guess I'd need to explain a bit the ordering here. This is indeed opinionated :-) and is very much up to a discussion. 

**First phase**: once these are all done, `cupyx.interpolate` can be declared MVP: most common usage is possible. 
The first phase work can be done partly asynchronously:

- `make_interp_spline` is a factory which constructs BSpline instances from data.
- PPoly is a base class (similar to BSpline) needed for actual interpolators. `CubicHermiteSpline` is a `PPoly` subclass, which is then the base for pchip, akima and CubicSpline. 
The hard part is I guess PPoly, as it has cython code / needs C kernels. The rest is pure python.
- `CubicSpline` is listed in the 2nd phase mainly because its functionality is a subset of `make_interp_spline`. Since the work on BSpline is already quite far, I don't see a reason to not finish it up. (Did I mention that the list is opinionated?)
- RGI is independent
- RBF is independent

So the first phase seems to have four lines of work which can be done in parallel: BSpline/make_interp_spline, PPoly and subclasses, RGI, RBF.
In principle, once can pospone the BSpline work and concentrate on PPoly instead if there is a strong feel for it.

The **second phase**:

There is a small typo (which propagates from the original where I made this typo): BPoly, NdPPoly, CubicSpline and interp1d are not Qhull-based.

BPoly is very similar to PPoly but less used (Mainly superseded by CubicHermiteSpline), NdPPoly is underused, so can be left for later. interp1d is legacy.

All these can be pushed to later, after the QHull based stuff is done. And *that* needs qhull or a cuda analog. Is there a triangulation library for CUDA we can depend on? 

The last big question is what to do with FITPACK based smoothing splines. Hopefully this question can be postponed :-).

**Comment by andfoy** [comment]
Thanks @ev-br for the clarifications! I'll reflect them on the main roadmap.

Regarding QHull, I found the following resources:
 https://www.sciencedirect.com/science/article/abs/pii/S0097849312000350

https://arxiv.org/abs/1508.05488

I haven't read them, but some SO answers point to them, that means that there is not a dedicated library to compute convex hulls in CUDA.

**Comment by ev-br** [comment]
One more big scipy dependency: RBFInterpolator depends on KDTree. cf gh-7190

**Comment by ideasrule** [comment]
I was incorporating cupy into my Python package about 2 weeks ago, and got stuck on the interpolation portions.  I'm relieved to see that you're working on this!

> In principle, once can pospone the BSpline work and concentrate on PPoly instead if there is a strong feel for it.

IMO having any non-linear interpolation function out there would be a vast improvement, and since the BSpline work is already pretty advanced, I think it should take priority.  (Disclaimer: I'm biased because I'm waiting to use the BSpline interpolation.)

If I wanted to contribute to this work, how would I do it?  Do I just pick something from the list without a check mark or a link to a pull request, and start working on it?

**Comment by leofang** [comment]
I am biased too and agree the most important ones are included in the first phase, great job in planning & execution!

**Comment by rgommers** [comment]
> If I wanted to contribute to this work, how would I do it? Do I just pick something from the list without a check mark or a link to a pull request, and start working on it?

You can just pick something and start digging in, CuPy welcomes PRs from everyone and it doesn't require pre-assigning a topic. However, given that this tracker is currently quite actively being worked on, I would suggest that once you've looked into a particular API for a little bit and think it's a good fit for you, to post here that you're tackling it - just to avoid double work.

**Comment by ideasrule** [comment]
I can work on interpn and RegularGridInterpolator (methods "linear" and "nearest").  Are there any guidelines on performance?  I can naively implement it by switching out numpy with cupy everywhere and changing a few things, but the performance might not be great.

**Comment by ev-br** [comment]
I think this would be very welcome @ideasrule!  

Not sure on the performance target, this is something we'll need to figure out indeed. That said, porting the scipy version to cupy would be  a reasonable first step! 
Once there is a port with assured correctness (tests!), we can work on performance. 

I'm pretty sure that in the end it'll need a dedicated kernel for finding the indices, similar to that of gh-7086, with an additional complication of having to deal with a tuple of dimensions, but *that* can be worked on when the base is there.

When working on this, best start from the scipy main version, it has been enhanced/refactored recently, cf https://github.com/scipy/scipy/blob/main/scipy/interpolate/_rgi.py

**Comment by ideasrule** [comment]
@ev-br I created a pull request of the naive implementation, together with passing unit tests copied from scipy.interpolate.  The pull request is #7197.  

Next, I'll benchmark the code with big arrays and see what the performance is.  If it's slow, I'll take a look at gh-7086 and see if I can write a kernel.

**Comment by andfoy** [comment]
@ideasrule, thanks a lot for your contribution! I'll add it to the tracker

**Comment by takagi** [comment]
All of the first phase items have been marked as done, thanks!

**Comment by takagi** [comment]
I'm seeing how it would be feasible to implement APIs categorized as problematic. We'd like to have equivalent GPU algorithms to those used in SciPy, which include Delaunay triangulation from QHull and kd tree that SciPy has its own implementation. I'm also looking into FITPACK, https://netlib.org/dierckx/.

**Comment by ev-br** [comment]
At least for FITPACK, I think the only  sane approach could be to rewrite the Fortran implementations first as a python prototype to break the monolith of the original Dierckx package.  It does not need to be python, as long as it's single-treaded on CPU. On a positive side, SciPy only wraps parts of the full Dierckx package. And the main missing piece is to figure out the strategy for inserting knots into the spline --- this is specific to P. Dierckx original package --- mainly it's `fpcurf.f`.  The rest we have, both both scipy and cupyx.scipy, thanks to Edgar's wrapping BSpline. Still quite a job.

An alternative is to port to CuPy `make_smoothing_spline` (new in scipy 1.10). It uses a different algorithm, is way less sophisticated ATM and is much less tested IRL.

**Comment by mullenkamp** [comment]
What's the closest function in CuPy to the LinearNDInterpolator for N-dimensions?
Would it be the RBFInterpolator using the 'linear' kernel?
Thanks!

**Comment by mullenkamp** [comment]
By the way, [CGAL](https://doc.cgal.org/latest/Manual/packages.html#PartTriangulationsAndDelaunayTriangulations) (a C++ library) may help with Delaunay Triangulations for the LinearNDInterpolator.

**Comment by ev-br** [comment]
lgpl licensing is incompatible, I'm afraid

**Comment by ev-br** [comment]
> What's the closest function in CuPy to the LinearNDInterpolator for N-dimensions?
Would it be the RBFInterpolator using the 'linear' kernel?

From what is implemented at the moment, yes, RBF is closest is spirit. 
The algorithm is quite dufferent, so results may differ. 

Implementing qhull-based interpolators is indeed on the TODO list, but it's a rather large project. If you know a license-compatible implementation, we'd love to hear about them!

**Comment by mullenkamp** [comment]
Thanks for the reply. And yes, I've tested both RBF and LinearNDInterpolator, and they do result in significant differences. Also LinearNDInterpolator is much faster.

Here are a couple other project I found related to Delaunay Triangulation for C++.

delaunator-cpp (only for 2D triangulation):
https://github.com/delfrrr/delaunator-cpp

CDT (may also only be for 2D...but it's not clear):
https://github.com/artem-ogre/CDT

OpenCV also has a 2D implementation:
https://stackoverflow.com/questions/11170564/is-there-c-api-for-delaunay-triangulation-in-opencv

Qhull has a C++ interface...but I'm guessing that doesn't help us:
https://github.com/qhull/qhull/wiki

It's really hard to find a C++ implementation for > 2D...

**Comment by ev-br** [comment]
Thank you for the list! 

> Also LinearNDInterpolator is much faster <than RBF>

https://github.com/cupy/cupy/pull/7864 should bring a perf improvement (at least, way better scaling with the dataset size).
